spreadsheet-weather-based-marketing/main-any-api.gs

```diff
@@ -0,0 +1,153 @@
+  // the trigger info object as a first param.
+  if (typeof onlyCheckAPI !== "boolean") {
+    onlyCheckAPI = false;
+  }
+
+  // Get items from Sheet
+  const rows = sheetsApi.get(configSpreadsheetName);
+
+  // Process sheet headers
+  config.setHeaders(rows[0]);
+  const apiHeaders = config.getApiHeaders();
+
+  // Configure all wrapper classes
+  const auth     = new Auth(config.get('service-account'));
+  const dv360    = new DV360(auth.getAuthToken());
+
+  sheetsApi.getSheetObject();
+
+  for (let i = 1; i < rows.length; i++) {
+    const row = rows[i];
+    const iPlus1 = i + 1;
+    
+    // Check if we already processed this item
+    const currentDateTime = new Date();
+    const lastUpdated = new Date(
+      row[ config.getHeaderIndex('col-last-updated') ]
+    );
+    
+    const diffHours = (currentDateTime - lastUpdated) / 1000 / 60 / 60;
+    const hoursBetweenUpdates = parseInt(config.get('hours-between-updates'));
+    if (!onlyCheckAPI && hoursBetweenUpdates && diffHours < hoursBetweenUpdates) {
+      Logger.log(`Row #${i} was already processed ${diffHours}h ago, skipping`);
+      continue;
+    }
+
+    const lineItemId = parseInt(row[config.getHeaderIndex('col-line-item-id')]),
+          insertionOrderId = parseInt(
+            row[config.getHeaderIndex('col-insertion-order-id')]
+          ),
+          advertiserId = parseInt(
+            row[config.getHeaderIndex('col-advertiser-id')]
+          ),
+          lat = parseFloat(row[config.getHeaderIndex('col-lat')]),
+          lon = parseFloat(row[config.getHeaderIndex('col-lon')]);
+
+    // Get API response
+    const anyApi = new AnyApi(
+      row[ config.getHeaderIndex('any-api-url') ],
+      row[ config.getHeaderIndex('any-api-headers') ],
+      config.getHeaders()
+    );
+    const resp = anyApi.getWithParams(row);
+
+    // Extract all weather variables
+    for (apiHeader in apiHeaders) {
+      row[ apiHeaders[apiHeader] ] = Utils
+        .getValueFromJSON(apiHeader, resp);
+    }
+
+    if (!onlyCheckAPI) {
+      row[config.getHeaderIndex('col-last-updated')] = currentDateTime
+        .toISOString();
+    }
+
+    // Save weather conditions back to Sheet
+    if (!sheetsApi.write([row], configSpreadsheetName + '!A' + iPlus1)) {
+      Logger.log('Error updating Sheet, retrying in 30s');
+      Utilities.sleep(30000);
+      
+      // Decrement `i` so that it ends up the same in the next for-loop iteration
+      i--;
+
+      continue;
+    }
+    
+    // Process activation formula
+    const formulaIdx = config.getHeaderIndex('col-formula') + 1;
+    sheetsApi.forceFormulasEval(iPlus1, formulaIdx);
+    const activate = sheetsApi.getCellValue(iPlus1, formulaIdx);
+    
+    if (!onlyCheckAPI) {
+      try {
+        // Switch Status according to the activation formula value
+        if (!isNaN(lineItemId) && lineItemId > 0) {
+          dv360.switchLIStatus(advertiserId, lineItemId, activate);
```

Comment:
can this be bulked? i.e. do we have to make a single API call for each row?

Reply:
Unfortunately yes (https://developers.google.com/display-video/api/reference/rest/v1/advertisers.lineItems/patch)